test/extensions/transport_sockets/alts/tsi_frame_protector_test.cc

```diff
@@ -0,0 +1,102 @@
+#include "common/buffer/buffer_impl.h"
+
+#include "extensions/transport_sockets/alts/tsi_frame_protector.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "src/core/tsi/fake_transport_security.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+using testing::InSequence;
+using testing::Invoke;
+using testing::NiceMock;
+using testing::SaveArg;
+using testing::Test;
+using testing::_;
+using namespace std::string_literals;
+
+/**
+ * Test with fake frame protector. The protected frame header is 4 byte length (little endian,
+ * include header itself) and following the body.
+ */
+class TsiFrameProtectorTest : public Test {
+public:
+  TsiFrameProtectorTest() : frame_protector_({tsi_create_fake_frame_protector(nullptr)}) {}
+
+protected:
+  TsiFrameProtector frame_protector_;
+};
+
+TEST_F(TsiFrameProtectorTest, Protect) {
+  {
+    Buffer::OwnedImpl input, encrypted;
+    input.add("foo");
+
+    EXPECT_EQ(TSI_OK, frame_protector_.protect(input, encrypted));
+    EXPECT_EQ("\x07\0\0\0foo"s, encrypted.toString());
+  }
+
+  {
+    Buffer::OwnedImpl input, encrypted;
+    input.add("foo");
+
+    EXPECT_EQ(TSI_OK, frame_protector_.protect(input, encrypted));
+    EXPECT_EQ("\x07\0\0\0foo"s, encrypted.toString());
+
+    input.add("bar");
+    EXPECT_EQ(TSI_OK, frame_protector_.protect(input, encrypted));
+    EXPECT_EQ("\x07\0\0\0foo\x07\0\0\0bar"s, encrypted.toString());
+  }
+
+  {
+    Buffer::OwnedImpl input, encrypted;
+    input.add(std::string(20000, 'a'));
```

Comment:
If you have a const for protected_buffer_size, you can use a relative larger number here instead of hard coding 20000.

Reply:
It is a coincidence that the frame size generated by fake frame protector is max at 16K and the buffer size is 16K, the hard code here is not reflecting the buffer size.

Reply:
I think this test case is testing protect() with |input| size larger than the buffer. So this line can be something more general, like: input.add(std::string(BUFFER_SIZE + 3620, 'a')); Am I misunderstanding?

Reply:
The frame size of fake frame protector output is define [here](https://github.com/grpc/grpc/blob/master/src/core/tsi/fake_transport_security.cc#L36), it is coincidence that is same as BUFFER_SIZE, but the test should be larger than the frame size, it is not depends on BUFFER_SIZE.

Reply:
Why is it testing against frame size which is not our implementation? I thought you meant to test protect() with input larger than |protected_buffer| which I think is worth to add if not covered yet.

Reply:
I meant the expected output below (L63) is depends on frame size in fake frame protector, the input here is to exceed the BUFFER_SIZE and they are just happened to be same.

We don't have visibility of BUFFER_SIZE in test.cc so I'll have to define BUFFER_SIZE, `input.add(std::string(BUFFER_SIZE + 3620, 'a'))` won't work anyway.